src/rosdiscover/observer/observer.py

```diff
@@ -45,17 +48,78 @@ def for_container(cls,
         rsw = roswire.ROSWire()
         app = rsw.app(config.image, config.sources)
         instance = app.attach(container, require_description=True)
+        observer = cls._build_observer(app, config, instance)
+        observer._x_start_script = None
+        observer._own_instance = False
+        return observer
+
+    @classmethod
+    def _build_observer(
+        cls,
+        app: App,
+        config: Config,
+        instance: AppInstance):
         if app.description.distribution.ros == ROSVersion.ROS1:
             from .ros1 import ROS1Observer
             return ROS1Observer(instance, config)
         else:
             from .ros2 import ROS2Observer
             return ROS2Observer(instance, config)
 
+    @classmethod
```

Comment:
I would suggest turning this method into a contextmanager and then creating and killing the container inside this method. That way, Observer doesn't need `._x_start_script` or `._own_instance`.